### PR TITLE
Fixes NAG optimizer #15543

### LINKED ISSUE #15543: Bug in NAG Optimizer
Current implementation of NAG is wrong.

**See this PDF for detail in the equations. [nag.pdf](https://github.com/apache/incubator-mxnet/files/3394333/nag.pdf)**

FYI @eric-haibin-lin

### PR DESCRIPTION
**Pull request by anirudhacharya**
## Description ##
Fixes https://github.com/apache/incubator-mxnet/issues/15543

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- fix update rule

For review - @zhanghang1989 @apeforest @eric-haibin-lin

### CODE REVIEW

**Comment by zhanghang1989** [comment]
image

```
mom = state
mom[:] *= self.momentum
weight[:] += lr * self.momentum * mom
weight[:] -= lr * (1 + self.self.momentum) * grad
mom[:] -=  grad
```

**Comment by wkcn** [comment]
Hi @zhanghang1989 , is there any difference between `a -= b` and `a[:] -= b`?
`a[:] -= b` may call extra function `__getitem__`.
```python
import mxnet as mx
import time

T = 1000
N = 1000


while 1:
    ti = time.time()
    a = mx.nd.arange(N)
    for i in range(T):
        a += 1
    mx.nd.waitall()
    print('a += b: ', time.time() - ti)

    ti = time.time()
    a = mx.nd.arange(N)
    for i in range(T):
        a[:] += 1
    mx.nd.waitall()
    print('a[:] += b: ', time.time() - ti)
```

Output:
```
a += b:  0.06155872344970703
a[:] += b:  0.3492248058319092
a += b:  0.06215381622314453
a[:] += b:  0.30852508544921875
a += b:  0.07872796058654785
a[:] += b:  0.31493425369262695
a += b:  0.08103752136230469
a[:] += b:  0.3226127624511719
a += b:  0.05706977844238281
a[:] += b:  0.29704785346984863

```

**Comment by anirudhacharya** [comment]
> image
> 
> ```
> mom = state
> mom[:] *= self.momentum
> weight[:] += lr * self.momentum * mom
> weight[:] -= lr * (1 + self.self.momentum) * grad
> mom[:] -=  grad
> ```

@zhanghang1989 The update rule in this PR is the following -

```cpp
mom_data[i] = param_momentum*mom_data[i];
KERNEL_ASSIGN(out_data[i], req, weight_data[i]-mom_data[i]
                              +(param_momentum+1)*(mom_data[i]
                                -(param_lr*(param_rescale_grad*grad_data[i]+param_wd*weight_data[i]))));
```
this update rule is same as the following psuedocode - 
```
weight = (weight - momentum * mom) + (momentum+1)*(momentum * mom - lr*(grad + wd*weight))
```
which when simplified, translates to 
```
weight[:] += (momentum**2 * mom) - (momentum + 1) * lr * (grad + wd*weight)
```
Formula - 
image

( it is the same rule used in keras as well - https://stats.stackexchange.com/questions/179915/whats-the-difference-between-momentum-based-gradient-descent-and-nesterovs-acc)

File: tests/python/unittest/test_optimizer.py [hunk 1]
```diff
@@ -400,10 +400,10 @@ def update(self, index, weight, grad, state):
...
             else:
                 mom[:] *= self.momentum
-                mom[:] += grad32
-                mom[:] += wd * weight32
-                grad32[:] += self.momentum * mom
-                weight32[:] -= lr * grad32
+                weight32[:] -= self.momentum * mom[:]
+                grad32 += wd * weight32
+                grad32 *= lr
+                weight32[:] -= (self.momentum + 1) * grad32
```
Comment (zhanghang1989) [inline tests/python/unittest/test_optimizer.py]: Should have `mom[:] +=  grad` in the end.

File: tests/python/unittest/test_optimizer.py [hunk 2]
```diff
@@ -385,10 +385,10 @@ def update(self, index, weight, grad, state):
...
               mom = state
               mom[:] *= self.momentum
-              mom[:] += grad
-              mom[:] += wd * weight
-              grad[:] += self.momentum * mom
-              weight[:] -= lr * grad
+              weight[:] -= self.momentum * mom[:]
+              grad += wd * weight
+              grad *= lr
+              weight[:] -= (self.momentum + 1) * grad
```
Comment (zhanghang1989) [inline tests/python/unittest/test_optimizer.py]: Should have `mom[:] +=  grad` in the end.

File: src/operator/optimizer_op-inl.h [hunk 3]
```diff
@@ -1119,22 +1112,16 @@ struct MP_NAGMomKernel {
...
-                    *static_cast<float>(grad_data[i]), param_clip_gradient)
-                    + (param_wd*w);
-      w = w - param_lr*(param_momentum*mom_data[i]
-                        + mshadow_op::clip::Map(param_rescale_grad
-                        *static_cast<float>(grad_data[i]),
-                        param_clip_gradient));
+      mom_data[i] = param_momentum*mom_data[i];
+      w = w-mom_data[i]+(param_momentum+1)*(mom_data[i]-param_lr
+              *(mshadow_op::clip::Map(param_rescale_grad*static_cast<float>(grad_data[i]),
+                          param_clip_gradient)+(param_wd*w)));
```
Comment (zhanghang1989) [inline src/operator/optimizer_op-inl.h]: Update the momentum state in the end `mom[:] -=  grad`

File: src/operator/optimizer_op-inl.h [hunk 4]
```diff
@@ -1119,22 +1112,16 @@ struct MP_NAGMomKernel {
...
     } else {
-      mom_data[i] = param_momentum*mom_data[i]
-                    + param_rescale_grad*static_cast<float>(grad_data[i])
-                    + (param_wd*w);
-      w = w - param_lr*(param_momentum*mom_data[i]
-       + param_rescale_grad*static_cast<float>(grad_data[i]));
+      mom_data[i] = param_momentum*mom_data[i];
+      w = w-mom_data[i]+(param_momentum+1)*(mom_data[i]-param_lr
+              *(param_rescale_grad*static_cast<float>(grad_data[i])+(param_wd*w)));
       weight32[i] = w;
```
Comment (zhanghang1989) [inline src/operator/optimizer_op-inl.h]: same here

**Comment by anirudhacharya** [comment]
> The weight update is correct. Please fix the mentum update in the end.

yes, will change the momentum update state

File: tests/python/unittest/test_optimizer.py [hunk 5]
```diff
@@ -399,11 +395,7 @@ def update(self, index, weight, grad, state):
             if self.momentum == 0.0:
                 weight32[:] += -lr * (grad32 + wd * weight32)
             else:
-                mom[:] *= self.momentum
-                weight32[:] -= self.momentum * mom[:]
-                grad32 += wd * weight32
-                grad32 *= lr
-                weight32[:] -= (self.momentum + 1) * grad32
+                weight32[:] += (self.momentum**2 * mom) - lr*(self.momentum+1)*(grad32 + wd*weight32)
```
Comment (zhanghang1989) [inline tests/python/unittest/test_optimizer.py]: Should multiply `lr` for the first term. `weight32[:] += lr*(self.momentum**2 * mom) - lr*(self.momentum+1)*(grad32 + wd*weight32)`

**Comment by zhanghang1989** [comment]
> Hi @zhanghang1989 , is there any difference between `a -= b` and `a[:] -= b`?
> `a[:] -= b` may call extra function `__getitem__`.
> 
> ```python
> import mxnet as mx
> import time
> 
> T = 1000
> N = 1000
> 
> 
> while 1:
>     ti = time.time()
>     a = mx.nd.arange(N)
>     for i in range(T):
>         a += 1
>     mx.nd.waitall()
>     print('a += b: ', time.time() - ti)
> 
>     ti = time.time()
>     a = mx.nd.arange(N)
>     for i in range(T):
>         a[:] += 1
>     mx.nd.waitall()
>     print('a[:] += b: ', time.time() - ti)
> ```
> 
> Output:
> 
> ```
> a += b:  0.06155872344970703
> a[:] += b:  0.3492248058319092
> a += b:  0.06215381622314453
> a[:] += b:  0.30852508544921875
> a += b:  0.07872796058654785
> a[:] += b:  0.31493425369262695
> a += b:  0.08103752136230469
> a[:] += b:  0.3226127624511719
> a += b:  0.05706977844238281
> a[:] += b:  0.29704785346984863
> ```

I am not familiar with symbol API. Just write some pseudocode to show how NAG works :)

File: tests/python/unittest/test_optimizer.py [hunk 6]
```diff
@@ -396,6 +397,7 @@ def update(self, index, weight, grad, state):
                 weight32[:] += -lr * (grad32 + wd * weight32)
             else:
                 weight32[:] += (self.momentum**2 * mom) - lr*(self.momentum+1)*(grad32 + wd*weight32)
+                mom = (self.momentum*mom) - lr*(grad32 + wd*weight32)
```
Comment (zhanghang1989) [inline tests/python/unittest/test_optimizer.py]: `lr` should be outside of the momentum state for mxnet standard practice.

Comment (anirudhacharya) [reply tests/python/unittest/test_optimizer.py]: why? but that is not nesterov logic right?

image

Comment (zhanghang1989) [reply tests/python/unittest/test_optimizer.py]: Please see SGD implementation:
http://mxnet.incubator.apache.org/api/python/optimization/optimization.html?highlight=optimizer.sgd#mxnet.optimizer.SGD

Comment (anirudhacharya) [reply tests/python/unittest/test_optimizer.py]: the sgd unit test also has `lr` as part of `mom` update - https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_optimizer.py#L142. 

let me know if you still think this needs to change.

Comment (zhanghang1989) [reply tests/python/unittest/test_optimizer.py]: Thanks! You are right. The MXNet implementation is puting lr inside. The update should be:


image
image

I will check your implementation again. Thanks!

**Comment by eric-haibin-lin** [comment]
Thanks @zhanghang1989 and @anirudhacharya

**Comment by Vikas-kum** [comment]
@anirudhacharya perl gpu tests are failing : http://jenkins.mxnet-ci.amazon-ml.com/blue/rest/organizations/jenkins/pipelines/mxnet-validation/pipelines/unix-gpu/branches/master/runs/1029/nodes/304/steps/568/log/?start=0 , 
this is the test that ran on master branch commit - https://github.com/apache/incubator-mxnet/commit/85fe0b439adf587024457a5346e8e84929195985

**Pull request merged**